### PR TITLE
Add OpenBuildings dataset

### LINKED ISSUE #68: Open Buildings Dataset
The dataset contains 516M building detections, across an area of 19.4M km2 (64% of the African continent).

For each building in this dataset we include the polygon describing its footprint on the ground, a confidence score indicating how sure we are that this is a building, and a Plus Code corresponding to the centre of the building. There is no information about the type of building, its street address, or any details other than its geometry.

* https://sites.research.google/open-buildings/
* https://ai.googleblog.com/2021/07/mapping-africas-buildings-with.html
* https://arxiv.org/abs/2107.12283

### PR DESCRIPTION
**Pull request by nilsleh**
This PR closes #68 and adds the [OpenBuildings dataset](https://sites.research.google/open-buildings/#download).

Dataset features:
- more than 516 M building detections across the African continent
- confidence score per building detection
- google Plus code corresponding to polygon building center

Dataset format:
- building data comes in 135 csv.gz files
- meta data file tiles.json that has geometry of each of the 130 single tile
- 5 csv.gz files are not covered in the meta data file since they are very small I suppose (one entry)

Questions/Concerns:
- I could not find the source crs 
- Currently, loading an item can take fairly long because some csv.gz contain millions of buildings
- It is quiet possible that a BoundingBox query returns a hit on a file but still finds no shapes for this area because the tiles are so large. 

Thus, I would welcome suggestions on these points, as well as any other suggestions of course.

Plot example:
openBuildings

### CODE REVIEW

**Comment by adamjstewart** [comment]
> I could not find the source crs

Can you elaborate on this? If there is no CRS, then there's no way to make this a GeoDataset, right? You have to have some transform to map this dataset from one CRS to another in order to use it geospatially.

File: docs/api/datasets.rst [hunk 1]
```diff
@@ -67,6 +67,11 @@ National Agriculture Imagery Program (NAIP)
 
 .. autoclass:: NAIP
 
+OpenBuildings
+^^^^^^^^^^^^^
```
Comment (adamjstewart) [inline docs/api/datasets.rst:71]:
```suggestion
Open Buildings
^^^^^^^^^^^^^^
```

File: torchgeo/datasets/openbuildings.py [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+from torch import Tensor
+
+from .geo import VectorDataset
+from .utils import BoundingBox, check_integrity
+
+
+class OpenBuildings(VectorDataset, abc.ABC):
+    r"""OpenBuildings dataset.
+
+    The `OpenBuildings
```
Comment (adamjstewart) [inline torchgeo/datasets/openbuildings.py:32]:
```suggestion
    r"""Open Buildings dataset.

    The `Open Buildings
```

File: torchgeo/datasets/openbuildings.py [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+import torch
+from rasterio.crs import CRS
+from rtree.index import Index, Property
+from torch import Tensor
+
+from .geo import VectorDataset
+from .utils import BoundingBox, check_integrity
+
+
+class OpenBuildings(VectorDataset, abc.ABC):
```
Comment (adamjstewart) [inline torchgeo/datasets/openbuildings.py:29]:
```suggestion
class OpenBuildings(VectorDataset):
```
Not an ABC

File: torchgeo/datasets/openbuildings.py [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+            transformed geometry in geojson format
+
+        """
+        x = json.dumps(shapely.geometry.mapping(wkt.loads(x)))
+        x = json.loads(x.replace("'", '"'))
+        return fiona.transform.transform_geom(
+            self._source_crs.to_dict(), self.crs.to_dict(), x
+        )
+
+    def _verify(self) -> None:
```
Comment (adamjstewart) [inline torchgeo/datasets/openbuildings.py:388]: We should probably also verify that a `tiles.geojson` file exists

File: torchgeo/datasets/openbuildings.py [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+    def __getitem__(self, query: BoundingBox) -> Dict[str, Any]:
+        """Retrieve image/mask and metadata indexed by query.
+
+        Args:
+            query: (minx, maxx, miny, maxy, mint, maxt) coordinates to index
+
+        Returns:
+            sample of image/mask and metadata for the given query. If there are
+            not matching shapes found within the query, an empty raster is returned
+
```
Comment (adamjstewart) [inline torchgeo/datasets/openbuildings.py:300]: Can remove one of these empty lines here

File: torchgeo/datasets/openbuildings.py (hunk 4 above)
Comment (nilsleh) [reply torchgeo/datasets/openbuildings.py:388]: I think I am verifying that in line 408.

**Comment by nilsleh** [comment]
> Can you elaborate on this? If there is no CRS, then there's no way to make this a GeoDataset, right? You have to have some transform to map this dataset from one CRS to another in order to use it geospatially.

I could not find a crs in the metadata file or the csv files where the polygon strings are. However, I assumed that since there are geographical coordinates, there would also be a CRS. But since it is not the case, I was wondering if there is maybe another way to find out or to handle it?

**Comment by adamjstewart** [comment]
Hmm, if the coordinates range from [-90, 90] and [-180, 180] or [0, 360] then we can probably assume that things are simply in lat/long units?

File: torchgeo/datasets/openbuildings.py (hunk 4 above)
Comment (adamjstewart) [reply torchgeo/datasets/openbuildings.py:386]: Ah yes, sorry, missed that.

**Review by adamjstewart — APPROVED** [review]

File: torchgeo/datasets/openbuildings.py [hunk 6]
```diff
@@ -0,0 +1,462 @@
...
+import glob
+import json
+import os
+import sys
+from typing import Any, Callable, Dict, List, Optional
+
+import fiona
+import fiona.transform
+import matplotlib.pyplot as plt
+import pandas as pd
```
Comment (adamjstewart) [inline torchgeo/datasets/openbuildings.py:15]: This should be a lazy import, we don't yet have a required dependency on pandas. This is causing the RTD tests to fail. Not sure why the Windows tests are failing though. See `idtrees.py` for an example where we do this. The tests should also skip if pandas isn't found, see `test_idtrees.py`

**Review by adamjstewart — DISMISSED** [review]

**Review by adamjstewart — APPROVED** [review]

**Pull request merged**